### PR TITLE
[Feature] add training process of the reid model

### PR DESCRIPTION
**Pull request by ToumaKazusa3**
Hi~
This PR add the following features:
1. Refactor `forward_train` in `BaseReid` class and `LinearReIDHead` class.
2. Add `TripletLoss`
3. Add the unitest of the above.

### CODE REVIEW

File: mmtrack/models/losses/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+from .hard_mine_triplet_loss import TripletLoss
```
Comment (OceanPang) [inline mmtrack/models/losses/__init__.py:1]: Just name the file "triplet_loss.py", and add `hard_mining` as an argument

File: mmtrack/models/reid/linear_reid_head.py [hunk 2]
```diff
@@ -32,17 +38,38 @@ def __init__(self,
...
                  act_cfg=None,
                  num_classes=None,
-                 loss=dict(type='CrossEntropyLoss', loss_weight=1.0),
+                 loss_cls=dict(type='CrossEntropyLoss', loss_weight=1.0),
+                 loss_triplet=dict(
+                     type='TripletLoss', margin=0.3, loss_weight=1.0),
+                 cal_acc=False,
                  topk=(1, )):
-        super(LinearReIDHead, self).__init__(loss=loss, topk=topk)
+        super(LinearReIDHead, self).__init__(
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:46]: 1. If you want to do that, there should be `loss=loss_cls` in the init function, not the hard code. 
1. We can keep the original `loss` in the arguments, and add `loss_pairwise` for triplet loss

File: mmtrack/models/reid/linear_reid_head.py [hunk 3]
```diff
@@ -32,17 +38,38 @@ def __init__(self,
...
+            loss=dict(type='CrossEntropyLoss'), topk=topk)
         self.num_fcs = num_fcs
         self.in_channels = in_channels
         self.fc_channels = fc_channels
         self.out_channels = out_channels
         self.norm_cfg = norm_cfg
         self.act_cfg = act_cfg
         self.num_classes = num_classes
-        self.loss = build_loss(loss)
+        self.compute_accuracy = Accuracy(topk=self.topk)
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:55]: `self.accuracy` is enough

File: mmtrack/models/reid/linear_reid_head.py [hunk 4]
```diff
@@ -32,17 +38,38 @@ def __init__(self,
...
+        if not loss_cls:
+            if isinstance(num_classes, int):
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:57]: We do not need an argument to set whether calculate the acc, just cal it if we have ce loss

File: mmtrack/models/reid/linear_reid_head.py [hunk 5]
```diff
@@ -32,17 +38,38 @@ def __init__(self,
...
+            if cal_acc:
+                warnings.warn('Since cross entropy is not set, '
+                              'the cal_acc will be ignored.')
+            if not loss_triplet:
+                raise ValueError('Please choose at least one loss in '
+                                 'triplet loss and cross entropy loss.')
+        elif not isinstance(num_classes, int):
+            raise ValueError('The num_classes must be a current number, '
+                             'if there is cross entropy loss.')
+        self.compute_loss_cls = build_loss(loss_cls) if loss_cls else None
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:69]: `self.loss_cls`

File: mmtrack/models/reid/linear_reid_head.py [hunk 6]
```diff
@@ -32,17 +38,38 @@ def __init__(self,
...
+        self.compute_loss_triplet = build_loss(
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:70]: self.loss_pairwise

File: mmtrack/models/reid/linear_reid_head.py [hunk 7]
```diff
@@ -57,23 +84,47 @@ def _init_layers(self):
...
+            constant_init(self.bn, 1, bias=0)
+            normal_init(self.classifier, mean=0, std=0.01, bias=0)
 
     def simple_test(self, x):
         """Test without augmentation."""
         for m in self.fcs:
             x = m(x)
-        x = self.fc_out(x)
-        return x
+        fea = self.fc_out(x)
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:102]: `feats`

File: mmtrack/models/losses/triplet_loss.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import torch
+import torch.nn as nn
+from mmcls.models import LOSSES
```
Comment (OceanPang) [inline mmtrack/models/losses/triplet_loss.py:3]: Use the `LOSSES` in this codebase is better.

File: mmtrack/models/losses/triplet_loss.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        Hermans et al. In Defense of the Triplet Loss for
+            Person Re-Identification. arXiv:1703.07737.
+    Imported from `<https://github.com/KaiyangZhou/deep-person-reid/blob/
+        master/torchreid/losses/hard_mine_triplet_loss.py>`_.
+    Args:
+        margin (float, optional): margin for triplet. Default is 0.3.
+        loss_weight (float, optional): Weight of the loss. Defaults to 1.0.
+    """
+
+    def __init__(self, margin=0.3, loss_weight=1.0, method='hard_mining'):
```
Comment (OceanPang) [inline mmtrack/models/losses/triplet_loss.py:20]: `hard_mining=True`

File: mmtrack/models/reid/linear_reid_head.py [hunk 10]
```diff
@@ -20,8 +23,11 @@ class LinearReIDHead(ClsHead):
         act_cfg (dict, optional): Configuration of activation method after fc.
             Defaults to None.
         num_classes (int, optional): Number of the identities. Default to None.
-        loss (dict, optional): Loss to train the re-identificaiton module.
-        topk (int, optional): Calculate topk accuracy.
+        loss_cls (dict, optional): Cross entropy loss to train the
+            re-identificaiton module.
+        loss_triplet (dict, optional): Triplet loss to train the
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:28]: `loss_pairwise`

File: mmtrack/models/reid/linear_reid_head.py [hunk 11]
```diff
@@ -57,23 +86,45 @@ def _init_layers(self):
...
-        x = self.fcs(x)
-        x = self.fc_out(x)
-        if self.num_classes is not None:
-            raise NotImplementedError()
-        losses = self.loss(x, gt_label)
+        for m in self.fcs:
+            x = m(x)
+        feats = self.fc_out(x)
+        losses = dict()
+        if not self.loss_cls:
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:113]: It shouldn't be an if-else logic. We can support  two losses at the same time.

File: mmtrack/models/reid/linear_reid_head.py [hunk 12]
```diff
@@ -57,23 +86,45 @@ def _init_layers(self):
...
+            losses['loss'] = self.loss_triplet(feats, gt_label)
+        else:
+            fea_bn = self.bn(feats)
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:116]: do it in the inplace manner

File: mmtrack/models/losses/triplet_loss.py (hunk 8 above)
Comment (ToumaKazusa3) [reply mmtrack/models/losses/triplet_loss.py:3]: This codebase do not have `LOSSES`.

File: mmtrack/models/losses/triplet_loss.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+class TripletLoss(nn.Module):
+    """Triplet loss with hard positive/negative mining.
+
+    Reference:
+        Hermans et al. In Defense of the Triplet Loss for
+            Person Re-Identification. arXiv:1703.07737.
+    Imported from `<https://github.com/KaiyangZhou/deep-person-reid/blob/
+        master/torchreid/losses/hard_mine_triplet_loss.py>`_.
+    Args:
+        margin (float, optional): margin for triplet. Default is 0.3.
```
Comment (OceanPang) [inline mmtrack/models/losses/triplet_loss.py:16]: 1. The styles of the captions should be consistent. You can use `Default: ***`, or `default to **`
2. The descriptions should be completed words, at least it should be `margin for triplet loss` here

File: mmtrack/models/reid/linear_reid_head.py [hunk 14]
```diff
@@ -33,16 +39,39 @@ def __init__(self,
                  act_cfg=None,
                  num_classes=None,
                  loss=dict(type='CrossEntropyLoss', loss_weight=1.0),
+                 loss_pairwise=dict(
```
Comment (OceanPang) [inline mmtrack/models/reid/linear_reid_head.py:42]: The default losses should be all None, nor the users will be always using the loss if they ignore the argument

File: mmtrack/models/losses/triplet_loss.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            dist_ap.append(dist[i][mask[i]].max().unsqueeze(0))
+            dist_an.append(dist[i][mask[i] == 0].min().unsqueeze(0))
+        dist_ap = torch.cat(dist_ap)
+        dist_an = torch.cat(dist_an)
+
+        # Compute ranking hinge loss
+        y = torch.ones_like(dist_an)
+        return self.loss_weight * self.ranking_loss(dist_an, dist_ap, y)
+
+    def forward(self, inputs, targets, **kwargs):
```
Comment (OceanPang) [inline mmtrack/models/losses/triplet_loss.py:60]: If you do not use the `kwargs`, remove it.

File: mmtrack/models/reid/base_reid.py [hunk 16]
```diff
@@ -10,9 +10,14 @@ class BaseReID(ImageClassifier):
...
 
-    def forward_train(self, *args, **kwargs):
+    def forward_train(self, img, gt_label, **kwargs):
         """"Training forward function."""
-        raise NotImplementedError()
+        x = self.extract_feat(img)
+
+        losses = dict()
+        loss = self.head.forward_train(x, gt_label)
+        losses.update(loss)
```
Comment (OceanPang) [inline mmtrack/models/reid/base_reid.py:19]: This function is consistent with its parent class, we can directly remove it.

File: mmtrack/models/reid/base_reid.py [hunk 17]
```diff
@@ -10,11 +10,19 @@ class BaseReID(ImageClassifier):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-    def forward_train(self, *args, **kwargs):
+    def forward_train(self, img, gt_label, **kwargs):
         """"Training forward function."""
-        raise NotImplementedError()
+        if img.ndim == 5:
```
Comment (OceanPang) [inline mmtrack/models/reid/base_reid.py:15]: Add annotation here to say what do you want to do if the dim == 5

File: mmtrack/models/losses/triplet_loss.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    """Triplet loss with hard positive/negative mining.
+
+    Reference:
+        Hermans et al. In Defense of the Triplet Loss for
+            Person Re-Identification. arXiv:1703.07737.
+    Imported from `<https://github.com/KaiyangZhou/deep-person-reid/blob/
+        master/torchreid/losses/hard_mine_triplet_loss.py>`_.
+    Args:
+        margin (float, optional): Margin for triplet loss. Default to 0.3.
+        loss_weight (float, optional): Weight of the loss. Defaults to 1.0.
```
Comment (OceanPang) [inline mmtrack/models/losses/triplet_loss.py:17]: Why there are `Default` and `Defaults`

**Review by OceanPang — APPROVED** [review]
